misc/random.c

```diff
@@ -80,12 +73,23 @@ uint64_t mp_rand_next(void)
     state[2] ^= t;
     state[3] = rotl_u64(state[3], 45);
 
-    mp_mutex_unlock(&state_mutex);
-
     return result;
 }
 
-double mp_rand_next_double(void)
+double mp_rand_next_double(struct mp_rand_state *s)
+{
+    return (mp_rand_next(s) >> 11) * 0x1.0p-53;
+}
+
+// Ref: https://www.pcg-random.org/posts/bounded-rands.html#bitmask-with-rejection-unbiased-apples-method
+uint32_t mp_rand_in_range32(struct mp_rand_state *s, uint32_t min, uint32_t max)
 {
-    return (mp_rand_next() >> 11) * 0x1.0p-53;
+    mp_assert(min < max);
+    uint32_t range = max - min;
+    uint32_t mask = mp_round_next_power_of_2(range) - 1;
```

Comment:
Wanted to do 64 bit range, but there's no 64 bit clz or similar in the codebase. And 32 is enough for the playlist shuffle.

Reply:
You can define `mp_clzll`, also would be faster than doing all the fluff that `mp_round_next_power_of_2` is doing. You can take a look at https://github.com/FFmpeg/FFmpeg/blob/master/libavutil/intmath.h for inspiration.